src/types/mod.rs

```diff
@@ -1,73 +1,62 @@
 use itertools::Itertools;
+use std::cmp::Ordering;
 use std::collections::{HashMap, HashSet};
 use std::hash::{Hash, Hasher};
 use std::iter;
 
 use super::parser;
 
-/// Represents a human-assigned name.
-pub type Name_ = Option<String>;
-
 /// Represents a unique identity, a [DeBruijn Index].
 ///
 /// [DeBruijn Index]: https://en.wikipedia.org/wiki/De_Bruijn_index
 pub type DeBruijn = usize;
 
-/// Represents the number of arguments an [`Operator`] takes.
-///
-/// [`Operator`]: trait.Operator.html
-pub type Arity_ = usize;
-
 /// Represents a place in a [`Term`]
 ///
 /// [`Term`]: struct.Term.html
 pub type Place = Vec<usize>;
 
-/// A trait for things that act like symbols
-pub trait Symbol: Eq + Hash + Clone {
-    /// Each symbol must have a unique ID associated with it.
-    type ID;
-    /// Return a new `ID` distinct from any existing `ID`.
-    fn new_id(&[&Self]) -> Self::ID;
-    /// Return a human-readable representation of `self`.
-    fn show(&self) -> String;
-    /// Return the `ID` of `self`.
-    fn id(&self) -> Self::ID;
-}
-
-/// An interface for things with arities.
-pub trait Arity {
-    fn arity(&self) -> Arity_;
-}
-
-/// An interface for things with names.
-pub trait Name {
-    fn name(&self) -> Name_;
-}
-
 /// An ADT for Variables.
-pub trait Variable: Symbol {}
+pub trait Variable: Eq + Ord + Hash + Clone {
+    /// Construct a new `Variable` that is larger than the given `Variable`.
+    fn next(Option<&Self>) -> Self;
+    /// A human-readable representation of `self`.
```

Comment:
Hmm... Variables don't necessarily need to be ordered so long as we can always construct a unique one on demand. Wouldn't something like `new_distinct` make more sense here, too? Is there some reason to think those two have to be connected?

Reply:
This is true — the TRS book's formalism grants an ordering out of convenience, not necessity. I had `new_distinct` and decided not to push it because it was quite messy. I think I will put `new_distinct` back though. I will have it take an `Option<String>` as well.